linera-sdk/src/test/block.rs

```diff
@@ -158,73 +154,94 @@ impl BlockBuilder {
-        self.incoming_bundles.extend(
-            message_ids
-                .into_iter()
-                .map(|message_id| (message_id, MessageAction::Accept)),
-        );
-        self
+        self.with_messages_from_by_medium(
+            certificate,
+            &Medium::Channel(ChannelFullName {
+                application_id: GenericApplicationId::System,
+                name: channel.name(),
+            }),
+        )
     }
 
-    /// Receives incoming messages by specifying them directly.
-    ///
-    /// This is an internal method that bypasses the check to see if the messages are already
-    /// present in the inboxes of the microchain that owns this block.
-    pub(crate) fn with_raw_messages(
+    /// Receives all direct messages  that were sent to this chain by the given certificate.
+    pub fn with_messages_from(&mut self, certificate: &Certificate) -> &mut Self {
+        self.with_messages_from_by_medium(certificate, &Medium::Direct)
+    }
+
+    /// Receives all messages that were sent to this chain by the given certificate.
+    pub fn with_messages_from_by_medium(
         &mut self,
-        messages: impl IntoIterator<Item = IncomingBundle>,
+        certificate: &Certificate,
+        medium: &Medium,
     ) -> &mut Self {
-        self.block.incoming_bundles.extend(messages);
+        let origin = Origin {
+            sender: certificate.value().chain_id(),
+            medium: medium.clone(),
+        };
+        let bundles = certificate
+            .message_bundles_for(medium, self.block.chain_id)
+            .into_iter()
+            .flat_map(|bundle| {
+                bundle
+                    .messages
+                    .into_iter()
+                    .map(|(index, message)| IncomingBundle {
+                        origin: origin.clone(),
+                        event: Event {
+                            certificate_hash: certificate.hash(),
+                            height: certificate.value().height(),
+                            index,
+                            authenticated_signer: message.authenticated_signer,
+                            grant: message.grant,
+                            refund_grant_to: message.refund_grant_to,
+                            kind: message.kind,
+                            timestamp: certificate
+                                .value()
+                                .executed_block()
+                                .unwrap()
+                                .block
+                                .timestamp,
+                            message: message.message,
+                        },
+                        action: MessageAction::Accept,
+                    })
+            });
+        self.with_incoming_bundles(bundles)
+    }
+
+    /// Rejects the most recently added message bundle.
+    pub fn with_rejection(&mut self) -> &mut Self {
+        self.block.incoming_bundles.last_mut().unwrap().action = MessageAction::Reject;
```

Comment:
I think I'd prefer to have explicit rejections instead of making the API mutate the last entry

```suggestion
    /// Rejects the most recently added message bundle.
    pub fn with_rejection_of_mesages_from(&mut self, certificate: &Certificate) -> &mut Self {
        self.with_mesages_from(certificate);
        self.block.incoming_bundles.last_mut().unwrap().action = MessageAction::Reject;
```